adapters/src/main/java/com/squareup/moshi/adapters/PolymorphicJsonAdapterFactory.kt

```diff
@@ -0,0 +1,276 @@
+ *   }
+ * }
+ *```
+ *
+ * Left unconfigured, Moshi would incorrectly attempt to decode the hand object to the abstract
+ * `HandOfCards` interface. We configure it to use the appropriate subtype instead:
+ *
+ * ```
+ * Moshi moshi = new Moshi.Builder()
+ *   .add(PolymorphicJsonAdapterFactory.of(HandOfCards.class, "hand_type")
+ *   .withSubtype(BlackjackHand.class, "blackjack")
+ *   .withSubtype(HoldemHand.class, "holdem"))
+ *   .build();
+ * ```
+ *
+ * This class imposes strict requirements on its use:
+ *  * Base types may be classes or interfaces.
+ *  * Subtypes must encode as JSON objects.
+ *  * Type information must be in the encoded object. Each message must have a type label like
+ * `hand_type` whose value is a string like `blackjack` that identifies which type
+ * to use.
+ *  * Each type identifier must be unique.
+ *
+ * For best performance type information should be the first field in the object. Otherwise Moshi
+ * must reprocess the JSON stream once it knows the object's type.
+ *
+ * If an unknown subtype is encountered when decoding:
+ *  * If [.withDefaultValue] is used, then `defaultValue` will be returned.
+ *  * If [.withFallbackJsonAdapter] is used, then the `fallbackJsonAdapter.fromJson(reader)` result will be returned.
+ *  * Otherwise a [JsonDataException] will be thrown.
+ *
+ * If an unknown type is encountered when encoding:
+ *  * If [.withFallbackJsonAdapter] is used, then the `fallbackJsonAdapter.toJson(writer, value)` result will be returned.
+ *  * Otherwise a [IllegalArgumentException] will be thrown.
+ *
+ * If the same subtype has multiple labels the first one is used when encoding.
+ */
+public class PolymorphicJsonAdapterFactory<T> internal constructor(
+  private val baseType: Class<T>,
+  private val labelKey: String,
+  private val labels: List<String>,
+  private val subtypes: List<Type>,
+  private val fallbackJsonAdapter: JsonAdapter<Any>?
+) : Factory {
+  /** Returns a new factory that decodes instances of `subtype`. */
+  public fun withSubtype(subtype: Class<out T>, label: String): PolymorphicJsonAdapterFactory<T> {
+    require(!labels.contains(label)) { "Labels must be unique." }
+    val newLabels: MutableList<String> = ArrayList(
+      labels
+    )
+    newLabels.add(label)
+    val newSubtypes: MutableList<Type> = ArrayList(
+      subtypes
+    )
+    newSubtypes.add(subtype)
+    return PolymorphicJsonAdapterFactory(
+      baseType = baseType,
+      labelKey = labelKey,
+      labels = newLabels,
+      subtypes = newSubtypes,
+      fallbackJsonAdapter = fallbackJsonAdapter
+    )
+  }
+
+  /**
+   * Returns a new factory that with default to `fallbackJsonAdapter.fromJson(reader)` upon
+   * decoding of unrecognized labels.
+   *
+   * The [JsonReader] instance will not be automatically consumed, so make sure to consume
+   * it within your implementation of [JsonAdapter.fromJson]
+   */
+  public fun withFallbackJsonAdapter(
+    fallbackJsonAdapter: JsonAdapter<Any>
```

Comment:
Java version has `@Nullable`
```suggestion
    fallbackJsonAdapter: JsonAdapter<Any>?
```